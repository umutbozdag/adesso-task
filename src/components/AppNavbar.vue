<template>
  <nav class="navbar bg-black border-gray-200 mb-12">
    <div class="flex flex-wrap items-center justify-between mx-auto py-4">
      <a class="flex items-center">
        <span class="self-center text-2xl font-semibold text-white">App</span>
      </a>
      <ul
        class="font-medium flex rounded-lg bg-black flex-row md:space-x-8 rtl:space-x-reverse mt-0 border-0 border-gray-700"
      >
        <template v-for="route in routes" :key="route.name">
          <li class="navbar__item" v-if="route.showInNavbar">
            <router-link
              :to="{ name: route.name }"
              class="navbar__link rounded md:bg-transparent px-2 py-1 text-white"
              >{{ route.label }}</router-link
            >
          </li>
        </template>
      </ul>
    </div>
  </nav>
</template>

<script setup lang="ts">
import routes from '@/config/routes';
</script>

<style lang="scss" scoped>
.navbar__link {
  &:hover {
    color: $color-beige;
    transition: 0.2s ease;
  }
}

.router-link-exact-active {
  background-color: $active-route-bg;
  color: $color-beige;
}
</style>
