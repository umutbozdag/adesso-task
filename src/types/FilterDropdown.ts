export interface FilterDropdownOption {
  label: string;
  name: string;
}

export enum FilterDropdownEmitEnum {
  OPTION_SELECTED = 'option-selected'
}
