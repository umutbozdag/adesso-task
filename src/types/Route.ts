export enum RoutesNameEnum {
  Home = 'Home',
  Units = 'Units',
  UnitDetails = 'UnitDetails',
  NotFound = 'NotFound'
}

export enum RoutesPathEnum {
  Home = '/',
  Units = '/units',
  UnitDetails = '/units/:id',
  NotFound = '/:pathMatch(.*)*'
}

export interface RouteConfigItem {
  path: RoutesPathEnum;
  name: RoutesNameEnum;
  showInNavbar?: boolean;
  label?: string;
}
