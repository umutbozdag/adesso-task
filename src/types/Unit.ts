export interface IUnits {
  units: IUnit[];
}

export interface IUnit {
  id: number;
  name: string;
  description: string;
  expansion: UnitExpansionType;
  age: UnitAgeType;
  cost: IUnitCost | null;
  build_time?: number;
  reload_time?: number;
  attack_delay?: number;
  movement_rate?: number;
  line_of_sight: number;
  hit_points: number;
  range?: UnitRangeType;
  attack?: number;
  armor: string;
  accuracy?: string;
  attack_bonus?: string[];
  search_radius?: number;
  blast_radius?: number;
  armor_bonus?: string[];
}

export enum UnitAgeEnum {
  All = 'All',
  Feudal = 'Feudal',
  Castle = 'Castle',
  Imperial = 'Imperial',
  Dark = 'Dark'
}

export type UnitAgeType = Exclude<UnitAgeEnum, UnitAgeEnum.All>;

export interface IUnitCost {
  Wood?: number;
  Gold?: number;
  Food?: number;
}

export type UnitExpansionType = 'Age of Kings' | 'The Conquerors';

export type UnitRangeType = number | string;
