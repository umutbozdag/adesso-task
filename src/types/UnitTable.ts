export interface IUnitTableHeaderItem {
  id: string;
  label: string;
}
