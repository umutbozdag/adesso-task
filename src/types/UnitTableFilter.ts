export enum UnitFilterOptionEnum {
  All = 'All',
  Feudal = 'Feudal',
  Castle = 'Castle',
  Imperial = 'Imperial',
  Dark = 'Dark'
}

export interface ICostFilter {
  label: string;
  sliderValue: string;
  selected: boolean;
}

export enum CostFilterEnum {
  WOOD = 'wood',
  FOOD = 'food',
  GOLD = 'gold'
}

export type CostFilterConfigType = Record<CostFilterEnum, ICostFilter>;
