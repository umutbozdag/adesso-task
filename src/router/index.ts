import { RouteRecordRaw, createRouter, createWebHistory } from 'vue-router';
import HomeView from '@/views/HomeView.vue';
import routeConfigs from '@/config/routes';
import { RoutesNameEnum } from '@/types/Route';

export const routes: RouteRecordRaw[] = [
  {
    path: routeConfigs[RoutesNameEnum.Home].path,
    name: routeConfigs[RoutesNameEnum.Home].name,
    component: HomeView
  },
  {
    path: routeConfigs[RoutesNameEnum.Units].path,
    name: routeConfigs[RoutesNameEnum.Units].name,
    component: () => import(/* webpackChunkName: "Units" */ '@/views/UnitsView.vue')
  },
  {
    path: routeConfigs[RoutesNameEnum.UnitDetails].path,
    name: routeConfigs[RoutesNameEnum.UnitDetails].name,
    component: () => import(/* webpackChunkName: "UnitDetails" */ '@/views/UnitDetailsView.vue')
  },
  {
    path: routeConfigs[RoutesNameEnum.NotFound].path,
    component: () => import(/* webpackChunkName: "NotFound" */ '@/components/NotFound.vue')
  }
];

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes
});

export default router;
