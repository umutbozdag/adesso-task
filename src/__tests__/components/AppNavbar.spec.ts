import { VueWrapper, mount } from '@vue/test-utils';
import { expect, vi, it, describe, beforeAll } from 'vitest';
import router from '@/router';
import { RoutesNameEnum } from '@/types/Route';
import AppNavbar from '@/components/AppNavbar.vue';
import { ComponentPublicInstance } from 'vue';

describe('AppNavbar', () => {
  let wrapper: VueWrapper<ComponentPublicInstance & { [key: string]: any }>;

  beforeAll(() => {
    wrapper = mount(AppNavbar, {
      global: {
        plugins: [router]
      }
    });
  });
  it('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.html()).toContain('Home');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows routes correctly', async () => {
    expect(wrapper.find('.navbar__item').exists()).toBe(true);
    expect(wrapper.html()).toContain('Home');
    expect(wrapper.html()).toContain('Home');
    await wrapper.find('.navbar__link').trigger('click');
  });

  it('router push', async () => {
    const routerPushSpy = vi.spyOn(router, 'push');
    router.push({ name: RoutesNameEnum.Units });
    expect(routerPushSpy).toHaveBeenCalled();
  });
});
