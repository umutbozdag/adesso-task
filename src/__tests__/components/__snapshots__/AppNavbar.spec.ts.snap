// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppNavbar > renders correctly 1`] = `
"<nav data-v-2694e7c3=\\"\\" class=\\"navbar bg-black border-gray-200 mb-12\\">
  <div data-v-2694e7c3=\\"\\" class=\\"flex flex-wrap items-center justify-between mx-auto py-4\\"><a data-v-2694e7c3=\\"\\" class=\\"flex items-center\\"><span data-v-2694e7c3=\\"\\" class=\\"self-center text-2xl font-semibold text-white\\">App</span></a>
    <ul data-v-2694e7c3=\\"\\" class=\\"font-medium flex rounded-lg bg-black flex-row md:space-x-8 rtl:space-x-reverse mt-0 border-0 border-gray-700\\">
      <li data-v-2694e7c3=\\"\\" class=\\"navbar__item\\"><a data-v-2694e7c3=\\"\\" href=\\"/\\" class=\\"router-link-active router-link-exact-active navbar__link rounded md:bg-transparent px-2 py-1 text-white\\" aria-current=\\"page\\">Home</a></li>
      <li data-v-2694e7c3=\\"\\" class=\\"navbar__item\\"><a data-v-2694e7c3=\\"\\" href=\\"/units\\" class=\\"navbar__link rounded md:bg-transparent px-2 py-1 text-white\\">Units</a></li>
      <!--v-if-->
      <!--v-if-->
    </ul>
  </div>
</nav>"
`;
